tests/foreman/virtwho/utils.py

```diff
@@ -0,0 +1,185 @@
+    :param str activation_key: the activation key will be used to register.
+    :param str org: Which organization will be used to register.
+    :raises: VirtWhoError: If failed to register the system.
+    """
+    runcmd('subscription-manager unregister', system)
+    runcmd('subscription-manager clean', system)
+    runcmd('rpm -qa | grep katello-ca-consumer | xargs rpm -e |sort', system)
+    runcmd(
+        'rpm -ihv http://{}/pub/katello-ca-consumer-latest.noarch.rpm'
+        .format(settings.server.hostname), system
+    )
+    cmd = 'subscription-manager register --org={} '.format(org)
+    if activation_key is not None:
+        cmd += '--activationkey={}'.format(activation_key)
+    else:
+        cmd += '--username={} --password={}'.format(
+            settings.server.admin_username,
+            settings.server.admin_password)
+    ret, stdout = runcmd(cmd, system)
+    if ret == 0 or "system has been registered" in stdout:
+        return True
+    else:
+        raise VirtWhoError(
+            'Failed to register system: {}'.format(system)
+        )
+
+
+def virtwho_cleanup():
+    """Before running test cases, need to clean the environment.
+    Do the following:
+    1. stop virt-who service.
+    2. kill all the virt-who pid
+    3. clean rhsm.log message, make sure there is no old message exist.
+    4. clean all the configure files in /etc/virt-who.d/
+    """
+    runcmd("systemctl stop virt-who")
+    runcmd(
+        "ps -ef | grep virt-who -i | grep -v grep |"
+        "awk '{print $2}' | xargs -I {} kill -9 {}"
+    )
+    runcmd("rm -f /var/run/virt-who.pid")
+    runcmd("rm -rf /var/log/rhsm/*")
+    runcmd("rm -rf /etc/virt-who.d/*")
+
+
+def get_virtwho_status():
+    """Return the status of virt-who service, it will help us to know
+    the virt-who configure file is deployed or not.
+    """
+    ret, stdout = runcmd('systemctl status virt-who')
+    running_stauts = ['is running', 'Active: active (running)']
+    stopped_status = ['is stopped', 'Active: inactive (dead)']
+    if ret == 0 and any(key in stdout for key in running_stauts):
+        return 'running'
+    elif ret == 0 and any(key in stdout for key in stopped_status):
+        return 'stopped'
+    else:
+        return 'undefined'
+
+
+def get_rhsm_logfile():
+    """Return a dictionary by analysing rhsm.log, and then can know
+        the virt-who configure file is deployed as expected or not.
+
+    The following keys will be included:
+    :[error]: how many error message exist, the expected value is 0.
+    :[reporter_id]: it can be defined by virtwho configure.
+    :[interval_time]: it can be defined by virtwho configure.
+    :[hypervisor_id]: it can be defined by virtwho configure.
+    :[hypervisor_name]: get to know the hypervisor's hostname.
+    :[guest_id]: get to know the guest_id.
+    """
+    ret, stdout = runcmd('cat /var/log/rhsm/rhsm.log')
+    data = dict()
+    data['error'] = len(re.findall(r'\[.*ERROR.*\]', stdout))
```

Comment:
Do error messages contain anything that could be used to map them to specific reporter/guest?

I guess what I would like to see is list of dicts constructed from entire file, instead of simple dict. This way you won't have to remove log files, and this solution will be much more robust than it is now.

Reply:
Error message validation is necessary, in our test experience, even the json info is found in rhsm.log, but there are some error message in rhsm.log, if that, we think the deployment should be failed.